New eden and heap sizing logic in Balanced GC

cedrichansen opened the pull request:
This change introduces 2 different major components to balanced gc, which are working together.

---------Change 1: New heap sizing logic in balanced gc
The old heap sizing logic, relied primarily on free memory %, in relation to -Xminf and -Xmaxf to decide whether or not the heap should shrink/expand.
If free memory was within the acceptable boundary, the heap sizing logic would then look at gc cpu% to decide whether to expand/contract.

The new logic (which is included in this commit), aims to take a more hybrid approach, where TENURE free memory % and GMP cpu % are working together, to determine whether or not the heap needs to change.
The function ```calculateCurrentHybridHeapOverhead``` determines what the current "hybrid overhead" of the heap/gc is (blend of gc cpu % and free memory %).
If this "hybrid overhead" is above -Xmaxt (default to 5%), the heap will expand. Conversely, if the "hybrid overhead" is below -Xmint (2% default), the heap will shrink. 

--------Change 2: New eden sizing logic in balanced

The current logic for eden sizing is fairly simple - By default, eden is taken to be 25% of the heap - unless specified otherwise by -Xmn/-Xmns/-Xmnx

The new logic, aims to make use of a few heuristics to improve overall PGC overheaad (% of time being active relative to total time), while trying to respect the newly introduced, `tarokTargetMaxPauseTime`.
Eden sizing logic now adheres to the following high level concepts:

 1. If the heap is not fully expanded ( that is current heap size <= -Xsoftmx/-Xmx), then eden will increase by 5% if PGC overhead and pgc avg time blend, is above
   dnssExpectedTimeRatioMinimum (5% default). Conversely, eden will shrink by 5% if the hybrid blend of pgc overhead and pgc time, falls below dnssExpectedTimeRatioMaximum (2% default). When the heap is not fully expanded, any changes in eden size, will result in the total heap to be resized by the same amount, in order to keep non-eden size the same.

 2. If the heap is fully expanded, a prediction will be made as to which eden size will produce the lowest total GC cpu overhead, while attempting to respect tarokTargetMaxPauseTime, given the constraints imposed by the heap (free memory, GMP cost, etc.
    This prediction accounts for PGC average time/intervals, GMP time/intervals, free Tenure space, and survival ratio of objects being copied out of eden, in order to determine the "best" eden size.
When the heap is fully expanded, and eden grows, it will "steal" free regions from "tenure"/non-eden.

Note: The following command line options are now supported by balanced gc policy as part of this change.
	- `-Xgc:dnssExpectedTimeRatioMinimum` -> sets min expected % of time spent in pgc pauses (default 2% for balanced)
	- `-Xgc:dnssExpectedTimeRatioMaximum` -> sets max expected % of time spent in pgc pauses (default 5% for balanced)
	- `-Xgc:targetPausetime` -> sets target (tarokTargetMaxPauseTime defaults to 200ms in balanced)

Depends on: eclipse/omr#5825

Signed-off-by: Cedric Hansen <cedric.hansen@ibm.com>

cedrichansen commented:
@amicic FYI

amicic commented on runtime/gc_vlhgc/MemorySubSpaceTarok.cpp:
```diff
@@ -952,12 +954,233 @@ void
+		sizeChange = calculateContractionSize(env, allocDescription, _systemGC, true);
+	}
+
+	if (0 == sizeChange && hybridHeapScore >= (double)_extensions->heapContractionGCTimeThreshold) {
```
As per coding standards, we put extra oval brackets and don't rely on operator priorities:

((0 == sizeChange) && (hybridHeapScore >= (double)_extensions->heapContractionGCTimeThreshold))

Please go through all the changes and add them when necessary. For this pass I'll ignore the rest of missing ones.

amicic commented:
One general  comment: if you think you touched a significant part of a file (50% would suffice), then convert J9 specific integer types like UDATA, I_32 etc... into standard types uintptr_t, int32_t etc in the whole file.

amicic commented:
Please create an Issue, and move most of PR/commit comments there.

cedrichansen commented:
This PR addresses https://github.com/eclipse/openj9/issues/12054

cedrichansen commented:
The latest 2 commits include the following changes:
- Fixed some "if" statements conditions to adhere to coding standards
- Changed j9 specific integer types, into standard integer types for files which saw significant modifications

dmitripivkine commented on runtime/gc_base/GCExtensions.cpp:
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 1991, 2020 IBM Corp. and others
+ * Copyright (c) 1991, 2021 IBM Corp. and others
```
This file after all has not been changed. Would you please just exclude it from change set without modification of copyrights data?

cedrichansen replied:
Done!

dmitripivkine commented on runtime/gc_modron_startup/mmparseXgc.cpp:
```diff
@@ -172,18 +180,20 @@ j9gc_initialize_parse_gc_colon(J9JavaVM *javaVM, char **scan_start)
 
-		goto _exit;
-	}
+		#if defined(J9VM_GC_REALTIME)
```
all preprocessor lines should start column 0

dmitripivkine commented on runtime/gc_vlhgc/GlobalMarkingScheme.cpp:
```diff
@@ -157,6 +157,14 @@ MM_ParallelGlobalMarkTask::setup(MM_EnvironmentBase *envBase)
+
+	/* Get gc threads cpu time for when mark started, and add it to the stats structure.
+	 * If the current platform does not support this operation, there are failsafes in place - no need for an else condition here
+	 */
```
please start multiline comments with `/*` in separate line:
```
        /* 
         * Get gc threads cpu time for when mark started, and add it to the stats structure.
 	 * If the current platform does not support this operation, there are failsafes in place - no need for an else condition here
 	 */
```

dmitripivkine commented on runtime/gc_vlhgc/MemorySubSpaceTarok.hpp:
```diff
@@ -60,6 +60,8 @@ class MM_MemorySubSpaceTarok : public MM_MemorySubSpace
 	MM_HeapRegionManager *_heapRegionManager;	/**< Stored so that we can resolve the table descriptor for given addresses when asked for a pool corresponding to a specific address */
 	MM_LightweightNonReentrantLock _expandLock; /**< Most of the common expand code is not multi-threaded safe (since it used in standard collectors on alloc path fail path which is single threaded)  */
+	double _lastObservedGcPercentage; /**< The most recently observed GC percentage (time gc is active / time gc is not active) */
+	U_64 _previouslyObservedPGCCount; /**< The count of observed PGC's since the last GMP cycle - Used to determine if heap resize occured after a system/global gc */
```
if you change variable types to standard in correspondent `.cpp` file, so it make sense to change types in `.hpp` as well as a pair

cedrichansen commented:
@dmitripivkine PR comments thus far have been addressed. These includes:
- Fixing multiline comments to adhere to style standards
- move preprocessor instruction to start of the line
- making `MemorySubspaceTarok.hpp` integer types consistent with .cpp integer types
- removed copyright update for untouched file  -> `GCextensionsBase.cpp`

amicic commented on runtime/gc_vlhgc/MemorySubSpaceTarok.cpp:
```diff
@@ -951,21 +953,245 @@ MM_MemorySubSpaceTarok::performResize(MM_EnvironmentBase *env, MM_AllocateDescri
+}	
+
+double
+MM_MemorySubSpaceTarok::calculateHybridHeapOverhead(MM_EnvironmentBase *env, IDATA heapChange)
```
there are a few of IDATAs that should be converted to intptr_t

amicic commented on runtime/gc_vlhgc/MemorySubSpaceTarok.cpp:
```diff
@@ -951,21 +953,245 @@ MM_MemorySubSpaceTarok::performResize(MM_EnvironmentBase *env, MM_AllocateDescri
+		Trc_MM_MemorySubSpaceTarok_calculateHybridHeapOverhead(env->getLanguageVMThread(), gcPercentage, freeMemComponant);
+	}
+
+	return 	(gcPercentage * gcOverheadWeight) + (freeMemComponant * (1 - gcOverheadWeight)); 
```
you can still use weightedAverage here

amicic commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -183,12 +261,32 @@ MM_SchedulingDelegate::globalGarbageCollectCompleted(MM_EnvironmentVLHGC *env, U
+	}
+	
+	double recentOverhead = (double)(_historicalPartialGCTime * 1000)/_averagePgcInterval;
+	_partialGcOverhead = (double)MM_Math::weightedAverage((float)_partialGcOverhead, (float)recentOverhead, 0.5);
```
introduce weightedAverage for double

cedrichansen replied:
https://github.com/eclipse/omr/pull/5825/commits/0aa8abb0f8331fe08856fbae27c8ba1dc22bb696 adds MM_Math::weightedAverage for doubles

amicic commented on runtime/gc_vlhgc/MemorySubSpaceTarok.cpp:
```diff
@@ -1355,18 +1385,169 @@ MM_MemorySubSpaceTarok::calculateExpandSize(MM_EnvironmentBase *env, UDATA bytes
+double 
+MM_MemorySubSpaceTarok::calculateGcPctForHeapChange(MM_EnvironmentBase *env, IDATA heapSizeChange)
+{
+	MM_EnvironmentVLHGC *envVLHGC = (MM_EnvironmentVLHGC *)env;
```
we typically name the one from the param envBase, and then this one just env

amicic replied:
You are right the existing pattern is predominant. I like what I proposed because the sites that use them will have a bit less characters, but it's not a strong argument. We can leave it as is - up to you.

amicic commented on runtime/gc_vlhgc/MemorySubSpaceTarok.cpp:
```diff
@@ -1355,18 +1385,169 @@ MM_MemorySubSpaceTarok::calculateExpandSize(MM_EnvironmentBase *env, UDATA bytes
+MM_MemorySubSpaceTarok::getHeapSizeWithinBounds(MM_EnvironmentBase *env)
+{
+	double currentHybridHeapScore = calculateCurrentHybridHeapOverhead(env);
+	uintptr_t reccomendedHeapSize = getActiveMemorySize();
```
spelling: recommended

cedrichansen commented:
@amicic PR comments have been addressed. These include:
- Converting `IDATA` -> `intptr_t` in `MemorySubSpaceTarok`
- Fix spelling mistake 'reccomended' -> 'recommended'
- using weightedAverage which takes `double` instead of `float`

I left out the change related to `envVLHGC` since some parts of the file in question use the same nomenclature I used for that same cast, but I can gladly change it if need be

cedrichansen commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -1157,16 +1652,42 @@ MM_SchedulingDelegate::updateGMPStats(MM_EnvironmentVLHGC *env)
+	if ((x1 < x2) && (y1 < y2)) {
+		double timeDiff = y1 - y2;
+		double edenSizeRatio = ((x1 + 1.0) / (x2 + 1.0));
+		_pgcTimeIncreasePerEdenRegionFactor = pow(edenSizeRatio, (1.0/timeDiff));
```
Just a note for reviewers - this `_pgcTimeIncreasePerEdenRegionFactor` **should not** be a weighted average.

Rationale - `_pgcTimeIncreasePerEdenRegionFactor` is not "linear" in the sense that the difference between 1.1 and 1.2, is not "the same" as 1.2 to 1.3, since this equation is finding the logarithmic fit between the 2 points. 
That being said, using a weighted average, is not really a good idea. Additionally, `_historicalPartialGCTime` is being used for this calculation, which is already a moving average, so no need to make `_pgcTimeIncreasePerEdenRegionFactor` a weighted average as well.

amicic replied:
This is probably ok for now, but this are still some assumed const here, which would fit for some workloads/current eden sizes and not for others. Specifically, A (assumed 1.0) here would could vary:

pow(edenSizeRatio, ( A /timeDiff));

How I would have done it (in theory anyway) is to remember last few (5-10) of (size, time) pairs  and do a linear regression. That would give you first derivative (slope of the whatever the real type of the curve is) in the _neighborhood_ of recent eden sizes. Indeed this is what we need, since we are typically doing incremental resizing (<10%), not that far from the recent past samples.

That approach would get rid of assumptions like minimumPgcTime, factor A etc, 
but it would be a lot more calculations and early estimates (during startup/rampup) would be fairly inaccurate.

cedrichansen replied:
Here the `1.0`, in `pow(edenSizeRatio, ( A /timeDiff));` is a workaround to represent  nthroot. The actual formula is the following 

![image](https://user-images.githubusercontent.com/28404014/111229238-f3c3b500-85bb-11eb-93ba-c78c28fccbc6.png)

Where `z` in the image, is  `_pgcTimeIncreasePerEdenRegionFactor` in the code. That being said, I don't really see the `1.0` changing, unless we change the formula (which might be an option).

As for the suggested improvement, I agree that this would be a good refinement, and I think we had briefly spoken about doing something along those lines, but as you mention, it adds more complexity/calculations.

amicic commented on runtime/gc_vlhgc/MemorySubSpaceTarok.cpp:
```diff
@@ -1355,18 +1384,169 @@ MM_MemorySubSpaceTarok::calculateExpandSize(MM_EnvironmentBase *env, UDATA bytes
+
+		if ((0 == pgcCount) && (0 == _lastObservedGcPercentage)) {
+			/* Very first time we are resizing, assume GC % is heapExpansionGCTimeThreshold */
+			_lastObservedGcPercentage = (double)_extensions->heapExpansionGCTimeThreshold;
```
would  (expansion + contraction)/2 be more reasonable?

cedrichansen replied:
My thought behind making this `heapExpansionGCTimeThreshold` is so that heap expansion would be "easier" to suggest at the very beginning of the run, although it likely doesn't matter too much - If free memory is very tight, the heuristic will suggest expansion anyways. Happy to change it to `(expansion + contraction)/2` if you prefer

amicic replied:
I leave it up to you - probably not much difference.

cedrichansen replied:
I'll keep it as `heapExpansionGCTimeThreshold `, but I will elaborate on the comment

amicic commented on runtime/gc_vlhgc/MemorySubSpaceTarok.cpp:
```diff
@@ -1355,18 +1384,169 @@ MM_MemorySubSpaceTarok::calculateExpandSize(MM_EnvironmentBase *env, UDATA bytes
+		 * The values below do not change if heapSizeChange != 0, since global GC is proportional to live data in heap, rather than heap size. 
+		 * This means changing the heap size will not change the gc% by much
+		 */
+		if(NULL != _collector) {		
```
```suggestion
		if (NULL != _collector) {		
```

amicic commented on runtime/gc_vlhgc/MemorySubSpaceTarok.cpp:
```diff
@@ -1355,18 +1384,169 @@ MM_MemorySubSpaceTarok::calculateExpandSize(MM_EnvironmentBase *env, UDATA bytes
+		if(NULL != _collector) {		
+			_lastObservedGcPercentage = (double)_collector->getGCTimePercentage(env);	
+		} else {	
+			_lastObservedGcPercentage= (double)_extensions->getGlobalCollector()->getGCTimePercentage(env);	
```
```suggestion
			_lastObservedGcPercentage = (double)_extensions->getGlobalCollector()->getGCTimePercentage(env);	
```

amicic replied:
in which case we wouldn't have _collector?

cedrichansen replied:
I'm not sure, however, the previous way to get GC percentage had this same condition (checking if there is a `_collector`), so I'm guessing whoever originally wrote that piece of code did so for a reason.

See `MM_MemorySubSpaceTarok::checkForRatioExpand()` in the changed files to see this usage.

amicic commented on runtime/gc_vlhgc/MemorySubSpaceTarok.cpp:
```diff
@@ -1355,18 +1384,169 @@ MM_MemorySubSpaceTarok::calculateExpandSize(MM_EnvironmentBase *env, UDATA bytes
+				 */
+				MM_EnvironmentVLHGC *envVLHGC = (MM_EnvironmentVLHGC *)env;
+				uintptr_t currentFreeTenure = (uintptr_t)envVLHGC->_heapSizingData.freeTenure;
+				uintptr_t potentialFreeTenure = currentFreeTenure + heapSizeChange;
+				if (heapSizeChange <= (-1 * (intptr_t)currentFreeTenure) ) {
+					/* If we try to shrink too much, too fast, tenure will be way too small, causing lots of GC work */
+					potentialFreeTenure = 1;
+				}
```
```suggestion
				uintptr_t potentialFreeTenure = 0;
				if (heapSizeChange <= (-1 * (intptr_t)currentFreeTenure) ) {
					/* If we try to shrink too much, too fast, tenure will be way too small, causing lots of GC work */
					potentialFreeTenure = 1;
				} else  {
					potentialFreeTenure = currentFreeTenure + heapSizeChange;
				}
```

amicic commented on runtime/gc_vlhgc/MemorySubSpaceTarok.cpp:
```diff
@@ -1355,18 +1384,169 @@ MM_MemorySubSpaceTarok::calculateExpandSize(MM_EnvironmentBase *env, UDATA bytes
+			}
+
+			double gcActiveTime = (double)(pgcCount * envVLHGC->_heapSizingData.avgPgcTimeUs) + envVLHGC->_heapSizingData.gmpTime; 
+			double gcInterval = (double)(pgcCount  * (envVLHGC->_heapSizingData.avgPgcTimeUs + envVLHGC->_heapSizingData.avgPgcIntervalUs)) + envVLHGC->_heapSizingData.gmpTime;
```
don't know what's meaning of gmpTime yet (is it CPU time spent, or wall clock time?)

I'm thinking that gcActiveTime should account for GMP CPU time (so that purely concurrent work would not be accouted, although up for debate), and gcInterval should account for GMP wall clock time

cedrichansen replied:
gmpTime is cpu time for GMP work (increments + sweep + concurrentWork/2), which does indeed mean that gcActiveTime includes GMP cpu time. 

I agree that the interval should be "wall clock" between start/end of GMP cycles. I have local variable called `globalMarkIntervalTime` in `SchedulingDelegate`, that I could pass over to be used instead of  `_heapSizingData.gmpTime`.
ie:
```
double gcInterval = (double)(pgcCount  * (envVLHGC->_heapSizingData.avgPgcTimeUs + envVLHGC->_heapSizingData.avgPgcIntervalUs)) + envVLHGC->_heapSizingData.gmpIntervalTime;
```

However, this will drastically change the gc % as reported by this function. If the interval between 2 gmp start (or end), is say 10s, and gc active time is 1s, then if we add GMP interval on the bottom (`gcInterval`), we effectively double the interval (10s becomes 20s), while the `gcActiveTime` remains the same. 

If we remove any GMP considerations in the bottom (`gcInterval`), then this is wall clock time between GMP starts - however, if GMP is very expensive, the gc % could be measured at over 100%. 
ex:
pgc count = 3,
pgc time = 20,
pgc interval = 200,
GMP workTime = 700

My original thought about keeping `envVLHGC->_heapSizingData.gmpTime` as part of the `gcInterval ` consideration, is that if `pgcCount` is very very low (say less than 3), then the measured GC percentage would be high, since GMP would constantly be active. On the other hand, if `pgcCount ` is very high, GMP work/time does not really matter into the overall GC %.

amicic replied:
I was thinking for gcInterval to use wall clock time _only_, without any PGC duration/interval addons. Sure if you start counting both, you'll end up with about double GMP interval values.

amicic commented on runtime/gc_vlhgc/IncrementalGenerationalGC.cpp:
```diff
@@ -1269,6 +1286,7 @@ MM_IncrementalGenerationalGC::partialGarbageCollectUsingCopyForward(MM_Environme
 
 	/* Record stats before a copy forward */
 	UDATA freeMemoryForSurvivor = _extensions->getHeap()->getActualFreeMemorySize();
+	env->_heapSizingData.freeTenure = freeMemoryForSurvivor;
```
In general, this should be reduced by eden size, right? 
But at this point eden is full, so this does happen to be correct? If so, put a comment above.

amicic replied:
Also should we reduce by reserved size (although it's likely done later at appropirate points when ths is used)

We should also try to avoid using work 'tenure' word in Balanced as much as possible, as it's already the case in the code (we don't have tenure on physical level), but if we do we should define variable (comment at declaration point) what we mean by that (for example, free memory excluding eden (and survivor reserved regions?)). 

Or should we just call it nonEden? It's already commonly used, although for live/used memory separation.

cedrichansen replied:
Exactly yes eden is full here. I will add a comment.

cedrichansen replied:
I like the idea of just adding comment at declaration points. I think if it's called `nonEden`, it might come across to a reader as not including survivor space. 

Maybe just calling it (albeit more verbose) `freeNonEdenAndSurvivor` - thoughts?

amicic commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -123,6 +142,58 @@ MM_SchedulingDelegate::getInitialTaxationThreshold(MM_EnvironmentVLHGC *env)
+
+	/* Determine how long it has been since previous global mark cycle started */
+	uint64_t globalMarkIntervalEndTime = j9time_hires_clock();
+	uint64_t globalMarkIntervalTime = j9time_hires_delta(_globalMarkIntervalStartTime, globalMarkIntervalEndTime, J9PORT_TIME_DELTA_IN_MICROSECONDS);
```
shouldn't this compare both end points (or both start points)?

cedrichansen replied:
It's comparing both start points. At the end of this very function, we have the following line

```
/* Set start time of next GMP phase, as end of current one */
_globalMarkIntervalStartTime = globalMarkIntervalEndTime;
```

amicic commented on runtime/gc_vlhgc/IncrementalGenerationalGC.cpp:
```diff
@@ -1203,8 +1212,15 @@ MM_IncrementalGenerationalGC::partialGarbageCollect(MM_EnvironmentVLHGC *env, MM
 		_reclaimDelegate.runGlobalSweepBeforePGC(env, allocDescription, env->_cycleState->_activeSubSpace, env->_cycleState->_gcCode);
 
+		/** Time indicated here is ~10% higher than that of corresponding sweeps timems in vgc logs - which is okay. It is factored into GMP overhead */
+		U_64 globalSweepTimeUs = j9time_hires_delta(globalSweepStartTime, j9time_hires_clock(), J9PORT_TIME_DELTA_IN_MICROSECONDS);
```
could we reuse start/end TS recorded by MM_ReclaimDelegate::doSweep() ?

amicic replied:
I don't see sweep duration being reported by trace (neither here, not later when you use it for GC overhead calculation). If so, please  expand Trc_MM_SweepEnd to report it (you'll have to obsolete the existing one and create V2)

amicic replied:
To minimize changes in OMR sweep code you could keep both version, and just make Balanced used the new one.

amicic commented on runtime/gc_vlhgc/IncrementalGenerationalGC.cpp:
```diff
@@ -1987,6 +2015,41 @@ MM_IncrementalGenerationalGC::postConcurrentUpdateStatsAndReport(MM_EnvironmentB
+
+		/* If there was a clock error, concurrentGCWorkTime might be too high or negative. Make sure ratio is between 0.1 and 0.8 */
+		concurrentGCRatio = OMR_MIN(concurrentGCRatio, 0.8);
+		concurrentGCRatio = OMR_MAX(concurrentGCRatio, 0.1);
```
Nice safety measures! Why not 0.9 (to be symmetric with 0.1)?
BTW, I'd like to reduce GC thread count in GMP increments, but it should be a separate change and experimentation.

cedrichansen replied:
No reason in particular. I can change to symmetrical values if you prefer, 0.1/0.9 or 0.2/0.8, let me know

amicic replied:
ok, let's flip a coin and say 0.1/0.9

amicic commented on runtime/gc_vlhgc/IncrementalGenerationalGC.cpp:
```diff
@@ -1928,6 +1946,13 @@ MM_IncrementalGenerationalGC::preConcurrentInitializeStatsAndReport(MM_Environme
 	stats->_scanTargetInBytes = _globalMarkPhaseIncrementBytesStillToScan;
 	env->_cycleState = &_persistentGlobalMarkPhaseState;
 	static_cast<MM_CycleStateVLHGC*>(env->_cycleState)->_vlhgcIncrementStats._markStats._startTime = j9time_hires_clock();
```
existing code, but we should store TS into a local and use it both to store into the stats and for the hook
same for post process method

cedrichansen replied:
Agreed, will refactor this portion.

amicic commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -1017,21 +1294,229 @@ MM_SchedulingDelegate::calculateEdenSize(MM_EnvironmentVLHGC *env)
+			 */
+			moveTowardRecommendedEden(env, 0.5);
+			resetPgcTimeStatistics(env);
+		} else if ((_pgcCountSinceGMPEnd & (consecutivePGCToChangeEden - 1)) == 0) {
```
id rather use arithmetic operations (modulo) and not rely on bitwise

amicic commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -1017,21 +1294,229 @@ MM_SchedulingDelegate::calculateEdenSize(MM_EnvironmentVLHGC *env)
+			/* Every consecutivePGCToChangeEden number of pgc's, re-evaluate eden size, and move towards it */
+			moveTowardRecommendedEden(env, 0.25);
+		}
+	} else if (0 == _pgcCountSinceGMPEnd % 3) {
```
I would be comfortable trying resize here on every PGC. It may be relevant during rampup phase. On the other side (we we have wide hysteresis bounds to prevent noise behavior at a stable state.

amicic replied:
BTW, scavenger in Gencon tries on each one.

The motivation for limiting attempts in fully expanded heap is to avoid recalculations for de-fragmenation work and GMP scheduling, although we can consider lifting the limitation there as well, at some point.

cedrichansen replied:
During testing/experimentation, I found that if we do this on every PGC, `_historicalPartialGCTime` doesn't have enough time to settle (as you point out). If PGC time is just slightly above pause time target, it can cause eden to decrease by an extra ~20% (ie, 3 decreases of 10%, rather than just 1 of 10%). 

Since `_historicalPartialGCTime` is historically weighed at 80%, it takes a while for values to settle, so my thought was that 3 pgc's would at least be halfway between old historical average, and new average - which would at least keep eden size from accidentally becoming too big/small too fast without an actual need.

amicic replied:
Then I'd consider reducing weight on old value (0.6 or 0.5) and let it learn new one faster. I'd rather deal with one param (only weight) than 2 (wieght and PGC skip count).
I don't see a problem getting 1 or even 2 extra adjustments due to inertia, as long as they still toward the CPU overhead optimum and not over shooting toward the other extreme. While CPU overhead bounds are rather wide, so overshoot should not occur that easily, I can that occurring for pause threshold (which has only one bound). Still it's a soft threshold, so we should be ok there, but you can play with the waight of pause threshold too to minimized those extra inertial adjustments.

amicic commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -1017,21 +1294,229 @@ MM_SchedulingDelegate::calculateEdenSize(MM_EnvironmentVLHGC *env)
+void
+MM_SchedulingDelegate::checkEdenSizeAfterPgc(MM_EnvironmentVLHGC *env, bool globalSweepHappened)
+{
+	if (!_extensions->statupPhaseFinished) {
```
What was your overall impression, how useful/important is to consider this case? (While fun learning exercise), for sake of simplicity, I'm willing to remove this.

cedrichansen replied:
From what I've seen in testing, in startup, GC times/overhead as well as survival rate (and a few other factors) are not representative of the actual regular allocation profile of the application (like at all...). That being said, making decisions on eden size during this time, might cause eden to shrink when it should expand, or vise versa.

amicic commented on runtime/gc_vlhgc/ConfigurationIncrementalGenerational.cpp:
```diff
@@ -287,6 +287,9 @@ MM_ConfigurationIncrementalGenerational::initialize(MM_EnvironmentBase *env)
 #define DEFAULT_MAX_AGE_FOR_PGC_COUNT_BASED				24
 #define DEFAULT_MAX_NURSERY_AGE							1
 #define DEFAULT_MAX_AGE_FOR_ALLOCATION_BASED			5
+#define DEFAULT_PGC_RATIO_FOR_EDEN_EXPANSION			0.07
```
in the light of future optimizations in PGC time, I'd like to optimistically reduce this to 6%

amicic commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -1017,21 +1294,229 @@ MM_SchedulingDelegate::calculateEdenSize(MM_EnvironmentVLHGC *env)
+	uintptr_t targetEdenBytes = currentIdealEdenBytes + targetEdenChange;
+	uintptr_t targetEdenRegions = targetEdenBytes / _regionManager->getRegionSize();
+
+	_edenSizeFactor = (intptr_t)targetEdenRegions - currentIdealEdenRegions;
```
For more consistent code at call site (checkEdenSizeAfterPgc), let's return the value and make the caller store it in _edenSizeFactor. The caller already deals with this factor in non-fully-expanded case.

cedrichansen replied:
100% agreed. Will make that change

amicic commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -1017,21 +1294,229 @@ MM_SchedulingDelegate::calculateEdenSize(MM_EnvironmentVLHGC *env)
+void
+MM_SchedulingDelegate::moveTowardRecommendedEden(MM_EnvironmentVLHGC *env, double edenChangeSpeed) 
+{
+	Assert_MM_true(edenChangeSpeed <= 1 && edenChangeSpeed >= 0);
```
extra oval brackets

amicic commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -1017,21 +1294,229 @@ MM_SchedulingDelegate::calculateEdenSize(MM_EnvironmentVLHGC *env)
+		return;
+	}
+
+	uintptr_t recommendedEdenSizeBytes = calculateRecommendedEdenSize(env);
```
While the code is very well documented, I would be shy to expand names of moveTowardRecommendedEden, calculateRecommendedEdenSize.... with ForExpandedHeap

amicic commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -523,44 +632,188 @@ MM_SchedulingDelegate::calculateEstimatedGlobalBytesToScan() const
+	/* Try "numberOfSamples" tests on the hybrid overhead curve, to determine which eden change will have best hybrid overhead */
+	for (uintptr_t i = 0; i < numberOfSamples; i++) {
+		/* Start from the right side of the curve */
+		intptr_t edenChange = (intptr_t)maxEdenChange - (samplingGranularity * i);
```
I'm missing something here - how do we evaluate for contraction (negative edenChange)!? Looks to me edenChange is always > minEdenChange (what in turn is > 0).

cedrichansen replied:
`minEdenChange` is simply equal to the negative of current eden size - meaning we are not allowed to try to shrink eden more than the actual size of eden. Consider the following.

5G heap, 1G live, 2G eden.

Eden can grow by at most be 2G (minus a little bit of room for survivor, more like ~1.5 G) -> `maxEdenChange `, and can be shrink by no more than 2G -> `minEdenChange` (which will always be negative).

This means that the change in eden size, is somewhere between -2G and 1.5G

amicic replied:
> minEdenChange is simply equal to the negative of current eden size
ah, I missed the negative part.

cedrichansen commented:
Latest commit addresses most PR comments, except the following, which is pending further discussion:
- Whether or not eden should be changed during startup phase
- If waiting for every 3 pgc's is necessary before resizing eden

amicic commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -523,44 +632,188 @@ MM_SchedulingDelegate::calculateEstimatedGlobalBytesToScan() const
+		} 
+	}
+
+	uintptr_t recommendedSize = currentIdealEdenSize + recommendedEdenChange;
```
Nitpick, but the method should be calculateRecommendedEdenChange... (not Size). You add currentIdealEdenSize here, and then at call site sub it.

cedrichansen replied:
Agreed, it's a bit redundant to add it in one spot, and then subtract in the other.

Will rename the function, as well as make the function simply return the change in eden bytes, rather than suggested eden size (along with fixing up this callsite)

amicic commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -523,44 +632,188 @@ MM_SchedulingDelegate::calculateEstimatedGlobalBytesToScan() const
+	}
+
+	uintptr_t recommendedSize = currentIdealEdenSize + recommendedEdenChange;
+	Trc_MM_SchedulingDelegate_calculateRecommendedEdenSize(env->getLanguageVMThread(), freeTenure, _totalGMPWorkTimeUs / 1000, avgPgcTimeUs, avgPgcIntervalUs, _edenSurvivalRateCopyForward, recommendedSize, bestOverheadPrediction);
```
trace points should reflect method renames

amicic commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -1017,21 +1294,229 @@ MM_SchedulingDelegate::calculateEdenSize(MM_EnvironmentVLHGC *env)
+	 * it wants to contract/expand, based on how much it will change the overhead and pgc times.
+	 */
+	double actualPGCOverheadWeight = 0.5;
+	assert(overhead >= 0.0 && overhead <= 1.0);
```
use MM_ style of assert and add extra oval brackets

amicic commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -1017,21 +1294,229 @@ MM_SchedulingDelegate::calculateEdenSize(MM_EnvironmentVLHGC *env)
+	double actualPGCOverheadWeight = 0.5;
+	assert(overhead >= 0.0 && overhead <= 1.0);
+	double pgcTimeOverhead = mapPgcTimeToPgcOverhead(env, partialGcTimeMs);
+	double hybridEdenOverheadHundredBased = (actualPGCOverheadWeight * (overhead * 100)) + ((1- actualPGCOverheadWeight) * pgcTimeOverhead);
```
you should still be able to use M_Math::weightedAverage (although we use it typically for historical weighted averaging)

amicic commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -1017,21 +1294,229 @@ MM_SchedulingDelegate::calculateEdenSize(MM_EnvironmentVLHGC *env)
+}
+
+double 
+MM_SchedulingDelegate::mapPgcTimeToPgcOverhead(MM_EnvironmentVLHGC *env, uintptr_t partialGcTimeMs) {
```
I'd rename to mapPgcPauseOverheadToPgcCPUOverhead. The key change was to use 'pause' instead of 'time' what slightly better represents what we are trying to optimize (that is meet pause threshold) - 'time' somewhat implies we are trying to optimize CPU overhead.

amicic replied:
consequnetly, I'd adjust parameter name and comments below to use 'pause'

amicic commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -1017,21 +1294,229 @@ MM_SchedulingDelegate::calculateEdenSize(MM_EnvironmentVLHGC *env)
+	double xmaxpct = _extensions->dnssExpectedTimeRatioMaximum._valueSpecified * 100;
+	double xmaxt = (double)_extensions->tarokTargetMaxPauseTime;
+
+	double overhead;
```
As per coding standards, this should be initied to a neutral value. Then, use can set it to midpointPct, and simplify one of the case down below.

amicic commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -1017,21 +1294,229 @@ MM_SchedulingDelegate::calculateEdenSize(MM_EnvironmentVLHGC *env)
+			 * If pgc time is only slightly above tarokTargetMaxPauseTime, then there is only a very small overhead penalty, 
+			 * wheras being 2x higher than the target pause time leads to a significantly bigger penalty  
+			 */
+			double overheadCurve = pow(1.03, ((double)partialGcTimeMs - xmaxt)) + midpointPct - 1;
```
This pow() will have very different behavior for potentially small xmaxt (1-10ms) vs large ones (100ms -1s). The point is, 2x violation should lead to the same overheadCurve value, regardless what the threshold is.

I'd like to see some normalization. Something like this. 

( base ^ ( partialGcTimeMs - xmaxt)/xmaxt  ) x midpointPct  + midpointPct 

'base' could be very different from what you used.

cedrichansen replied:
That makes sense, I'll fix that up

amicic commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -1017,21 +1294,229 @@ MM_SchedulingDelegate::calculateEdenSize(MM_EnvironmentVLHGC *env)
+		 */
+		double slope = (xmaxpct - xminpct) / ((xmaxt/2) - xmaxt);
+		overhead = (slope * (double)partialGcTimeMs) + ((2.0 * xmaxpct) - xminpct);
+		overhead = OMR_MAX(overhead, 0.0);	
```
if I calculated correctly
- with partialGcTimeMs >= xmaxt, this should return a value that is <= xminpct (hence explicitely suggesting contract)
- as partialGcTimeMs is approaching xmaxt/2, the return value is approaching xmaxpct (suggesting neither explicit contraction nor explicit expansion, although blended later with CPU overhead it may implicitly result into a contraction/expansion)
- with partialGcTimeMs  < xmaxt/2, the returned value is xmaxpct (still not explicitely suggesting expansion, but may quite help to occur)
It think it reasonable, although hard to quantify without trying some other mappings

If my comments are correct/sensible, you can consider embedding (and correcting/adjusting if needed) them into the code.

cedrichansen replied:
Yes thats correct.

One thing I have been thinking about, is whether or not `overhead = OMR_MAX(overhead, 0.0);` is desirable or not. If `partialGcTimeMs ` is far over the pause time target (whether it be 2x higher, or 100x higher), it will map to 0. If we allow this mapping to go negative, it will likely get a better balanced of pgc time & cpu overhead - despite a negative overhead not really making sense (from a high level point of view). Thoughts on removing the line which keeps the mapped overhead to >0?

amicic replied:
As said earlier, I can't tell how aggressively this mapping will suggest contraction (or expansion) without seeing it in action (and comparing with some other mapping). 
Being a new thing, let's play conservative (be fairly soft) in regards to the pause criteria, and keep the cap (keep it non-negative). Down the road (when we get some real life experience) we may consider some tuning.

amicic commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -1017,21 +1294,229 @@ MM_SchedulingDelegate::calculateEdenSize(MM_EnvironmentVLHGC *env)
+		overhead = OMR_MAX(overhead, 0.0);	
+		/*  Expanding simply because pgc time is very small is not a good idea, so return xmaxpct, so that if the pgc cpu overhead wants to expand, only then eden expands */
+		overhead = OMR_MIN(overhead, xmaxpct);
+	}
```
I wonder how much contradicting suggestions we might get if heap is close to being fully expanded and sometimes following one path and sometime the other. Ideally, we should not have a hard 'if expanded' check, but have a weighted blend of the 2 recommendations (based on how close to being fully expnded we are), but that would be more pain for future observers to understand. :0

cedrichansen replied:
Thats a good point. I have observed some small contradictions myself when heap size is fairly close to max size, although what I proposed [in this comment](https://github.com/eclipse/openj9/pull/12043#discussion_r593499193) would be one solution.

I think the solution you suggest (the blend, weighted on how close heap is to being fully expanded), might be another solution worth investigating a bit deeper (unless we choose to take the approach I mention in comment linked just above)

amicic replied:
Hm, to me those are 2 fairly different problems. If fact with letting overhead going negative you may even amplify the problem I'm trying to address. Very long GC will both affect CPU overhead criteria to try to expand and pause target criteria to try to contract (even more without capping), and with heap being sometimes fully expanded and sometimes just below, will potentially create large oscillations in eden sizing.

Let' give a try to blending....

cedrichansen commented:
PR feedback since my last commit has been addressed. The changes include:

- `calculateRecommendedEdenSizeForExpandedHeap()` -> `calculateRecommendedEdenChangeForExpandedHeap()` . This function now returns the recommended change in eden size, rather than the new recommended eden size. The caller of this function handles this change appropriatly.
- `mapPgcTimeToPgcOverhead()` -> `mapPgcPauseOverheadToPgcCPUOverhead()` . The increased verbosity better explains what the function does.
- Made a small modification to the way pgc pause time is mapped to cpu overhead when heap is fully expanded. The rationale for the change is outlined in the comments
- Renamed params in `calculateHybridEdenOverhead` from `partialGcTimeMs` -> `pgcPauseTimeMs` to better reflect the variables usage.
- changed an `assert` -> `Assert_MM_true`
- Modified a few comments to reflect changes mentioned above.

amicic commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -1017,21 +1293,230 @@ MM_SchedulingDelegate::calculateEdenSize(MM_EnvironmentVLHGC *env)
+	}
+	if (_extensions->userSpecifiedParameters._Xmn._wasSpecified || _extensions->userSpecifiedParameters._Xmnx._wasSpecified) {
+		maxEdenCount = _extensions->tarokIdealEdenMaximumBytes / _regionManager->getRegionSize();
+	}
```
I know this is cheap to recompute on each GC cycle, but I believe these 2 Xmn/Xmnx checks could be done once during SchedulingDelegate initialization and where would update _maxEdenPercent/_minEdenPercent

amicic commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -1058,74 +1549,80 @@ MM_SchedulingDelegate::getCurrentEdenSizeInRegions(MM_EnvironmentVLHGC *env)
 		edenIdealBytes = edenMaximumBytes;
 	} else {
 		/* interpolate between the maximum and minimum */
 		/* This logic follows the formula given in JAZZ 39694
```
you can remove this part of comment

amicic commented on runtime/gc_base/GCExtensions.hpp:
```diff
@@ -149,14 +149,17 @@ class MM_GCExtensions : public MM_GCExtensionsBase {
 #endif /*defined(J9VM_GC_DYNAMIC_CLASS_UNLOADING) */
 
-
+	bool statupPhaseFinished; /**< Used to indicate if the JIT startup phase is finished */
```
this is not used anymore - we should probably remove it?

cedrichansen replied:
Right... I'm not sure whether to keep it or not. On one hand, we don't currently use it, but on the other, I think this might be a useful flag for other scenarios in the future. Up to you if we keep it

amicic commented on runtime/gc_base/GCExtensions.hpp:
```diff
@@ -149,14 +149,17 @@ class MM_GCExtensions : public MM_GCExtensionsBase {
 	MM_Wildcard* numaCommonThreadClassNamePatterns; /**< A linked list of thread class names which should be associated with the common context */
 
+#if defined(J9VM_GC_VLHGC)
+	UDATA tarokTargetMaxPauseTime; /**< An optional, user specified soft max pause time for PGC's in balanced GC*/
```
is there a group of tarok specific parms already that you could move this to?

cedrichansen replied:
There aren't any VLHGC/tarok params elsewhere in this file

amicic replied:
ok, move then to OMR/Base super class, where we have lots of them
while it's really semantically less correct, at least we are consistently wrong for all those params

cedrichansen replied:
Moved it to `GCExtensionsBase` as per latest commit on OMR side https://github.com/eclipse/omr/pull/5825

amicic commented on runtime/gc_modron_startup/mmparseXgc.cpp:
```diff
@@ -172,18 +180,20 @@ j9gc_initialize_parse_gc_colon(J9JavaVM *javaVM, char **scan_start)
+#endif /* J9VM_GC_REALTIME */
+		
+#if defined(J9VM_GC_VLHGC)
+		    /* Save soft pause target for balanced */
```
try to group with existing #ifdef VLHGC

cedrichansen replied:
After discussion offline, leaving as is

amicic commented on runtime/gc_vlhgc/EnvironmentVLHGC.hpp:
```diff
@@ -79,6 +79,30 @@ class MM_EnvironmentVLHGC : public MM_EnvironmentBase
+		U_64 avgPgcIntervalUs;
+		U_64 pgcCountSinceGMPEnd;
+		U_64 reservedSize;
+		U_64 freeTenure;
```
use uint64_t and convert existing ones too (this is not a big file)

amicic commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -1017,21 +1318,240 @@ MM_SchedulingDelegate::calculateEdenSize(MM_EnvironmentVLHGC *env)
+	}
+
+	/** 
+	 * Once ratioOfHeapExpanded is over heapFullyExpandedThreshold, start to blend the recommendations provided by 
```
explain not only what we do it by why. something like: to ensure smooth transition of potentially very different recommended eden sizes for fully expanded case and non-expanded heap case when the cases frequently change from one to the other

cedrichansen replied:
Comment has been updated. See just a few lines below

cedrichansen commented:
The latest force push, includes a few small changes in `IncrementalGenerationalGC.cpp` after merge conflict due to https://github.com/eclipse/openj9/pull/12187

amicic commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -973,65 +1277,313 @@ MM_SchedulingDelegate::calculateAutomaticGMPIntermission(MM_EnvironmentVLHGC *en
+
+	double hybridEdenOverhead = calculateHybridEdenOverhead(env, (uintptr_t)_historicalPartialGCTime, _partialGcOverhead, false);
+
+	Trc_MM_SchedulingDelegate_calculateEdenChangeHeapNotFullyExpanded(env->getLanguageVMThread(), hybridEdenOverhead, (uintptr_t)_historicalPartialGCTime, mapPgcPauseOverheadToPgcCPUOverhead(env, (uintptr_t)_historicalPartialGCTime, false));
```
I'd move this trace point inside calculateHybridEdenOverhead (and rename it accordingly) to avoid double calculation of mapPgcPauseOverheadToPgcCPUOverhead. Then you can expand it to report heapFullyExpanded parameter, as well (you can 'convert' bool to strings easily with '?:' conditional operator) and make it useful in expanded case, too.

cedrichansen replied:
If I move this tracepoint inside of `calculateHybridEdenOverhead()`, then it will be printed 100 or so times. This is because `calculateRecommendedEdenChangeForExpandedHeap()` calls `calculateHybridEdenOverhead()` when it is trying to find the best value for eden size. 

That being said, I can add a parameter in `calculateHybridEdenOverhead()` which indicates if the tracepoint should be run.  I agree that recalculating `mapPgcPauseOverheadToPgcCPUOverhead ()` is a bit wasteful, but the alternative (passing in a parameter indicating whether or not the tracepoint should run), might be a little unorthodox (I haven't seen any other functions do this, but It certainly might exist somewhere). Up to you which approach is better.

genie-openj9 commented:
Can one of the admins verify this patch?

amicic commented on runtime/gc_stats/MarkVLHGCStats.hpp:
```diff
@@ -80,6 +80,9 @@ class MM_MarkVLHGCStats : public MM_MarkVLHGCStatsCore
 #endif /* J9MODRON_TGC_PARALLEL_STATISTICS */
 
+	U_64 _concurrentGCThreadsStartTimeSum; /**< The sum of all gc cpu thread times when concurrent gc work began */
+	U_64 _concurrentGCThreadsEndTimeSum; /**< The sum of all gc cpu thread times when concurrent gc work ended */
```
you could add CPU the names

cedrichansen replied:
renamed it to `_concurrentGCThreadsCPUStartTimeSum ` as per the latest commit.

amicic commented on runtime/gc_vlhgc/IncrementalGenerationalGC.cpp:
```diff
@@ -2018,6 +2061,41 @@ MM_IncrementalGenerationalGC::postConcurrentUpdateStatsAndReport(MM_EnvironmentB
+		/* If the platform supports getting time for each thread, calculate what % of time gc threads were active relative to application threads. */
+		/* Determine the ratio of time spent doing GC work vs non-gc related work during the concurrent phase */
+		U_64 concurrentGCWorkTime = markStats._concurrentGCThreadsCPUEndTimeSum - markStats._concurrentGCThreadsCPUStartTimeSum;
+		concurrentGCRatio = (double)concurrentGCWorkTime/concurrentElapsedTime;
```
hm, this ratio is a _sum_ of CPU from _multiple_ GC threads vs elapsed time. this is inflated number (by a number of threads). I guess this is typically > 1.0

cedrichansen replied:
We are comparing the sum of process time for GC threads to elapse total cpu time for all threads (application threads and GC threads). 

`concurrentElapsedTime ` is indeed very misleading. It should in fact be called, `concurrentElapsedCPUTime`, as per our discussion offline.

Renaming `concurrentGCWorkTime` to `concurrentGcCPUWorkTime` might also be a helpful touch for anyone looking at this code later on

cedrichansen replied:
Naming of these variables for increased clarity has been resolved as per latest commit

amicic commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -1157,16 +1721,43 @@ MM_SchedulingDelegate::updateGMPStats(MM_EnvironmentVLHGC *env)
+	if ((x1 < x2) && (y1 < y2)) {
+		double timeDiff = y1 - y2;
+		double edenSizeRatio = ((x1 + 1.0) / (x2 + 1.0));
+		_pgcTimeIncreasePerEdenFactor = pow(edenSizeRatio, (1.0/timeDiff));
```
We should report this value in a trace point. Perhaps you can piggyback on an existing one at caller site?

cedrichansen replied:
Agreed that this should be reported. 
As for which tracepoint to use, I don't see any tracepoints at caller site right now. I'll take a look at some of the other variables that get set at end of PGC, and include any metrics that currently aren't being reporting into this new tracepoint. If all variables are being reported, I'll just add a tracepoint in this function, and be a bit more verbose (ie, include `x1`, `x2`, `timeDiff`, etc)

dmitripivkine replied:
```suggestion
		_pgcTimeIncreasePerEdenFactor = pow(edenSizeRatio, (1.0 / timeDiff));
```

amicic commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -973,65 +1302,312 @@ MM_SchedulingDelegate::calculateAutomaticGMPIntermission(MM_EnvironmentVLHGC *en
+		/* Total heap needs to be aware that by changing eden size, the amount of survivor space might also need to change */
+		if (0 < desiredEdenChangeSize) {
+			edenChangeWithSurvivorHeadroom = desiredEdenChangeSize + (intptr_t)ceil(((double)desiredEdenChangeSize * _edenSurvivalRateCopyForward));
+		} else if ((0 > desiredEdenChangeSize) && (16 < _edenRegionCount)) {
```
There might be  (although rare) configurations where max regions counts is much smaller or bigger then default 1000-2000. Perhaps we can express this comparison with 16 in relative terms like:

_edenRegionCount * 64 > maxRegionCount 

Although the whole idea of special casing for startup with some hard coded ratio  representing 'small region' isn't ideal...

cedrichansen replied:
`_edenRegionCount * 64 > maxRegionCount` seems reasonable!

The only thing that this threshold is really doing, is including/excluding survivor regions when informing the total heap by how many regions it needs to change. Through experimentation, 16 was reasonable number of eden regions to start to include for the reduction of survivor regions needed, but the alternative you mentioned would also work just fine. Having some sort of case for when eden is very small (whether relative to maxRegions or some absolute value) just prevents the heap from shrinking more than it needs to

cedrichansen replied:
This has been fixed as of latest commit, and it behaves as expected. 
With very small heaps (tested with -Xmx32M) the condition allows eden to include survivor headroom essentially right away. (threshold basically becomes true when eden count is 1), but when 1000-2000 regions are used (the common case), this condition behaves the same was as before (anywhere from ~ 16-32 eden regions is now the threshold)

cedrichansen commented:
@amicic Both PR comments from Monday, July 26th have been addressed.

dmitripivkine commented on runtime/gc_modron_startup/mmparse.cpp:
```diff
@@ -919,24 +919,32 @@ gcParseSovereignArguments(J9JavaVM *vm)
-			j9nls_printf(PORTLIB, J9NLS_ERROR, J9NLS_GC_OPTIONS_PERCENT_OUT_OF_RANGE, "-Xmaxt", 0.0, 1.0);
+	
+	if (-1 != FIND_ARG_IN_VMARGS(EXACT_MEMORY_MATCH, "-Xmaxt", NULL)) {
+		extensions->heapExpansionGCRatioThreshold._wasSpecified = true;
```
I think conceptually it is make sense to set `_wasSpecified` for correctly formed (parsed) options only. Otherwise it creates "grey" area for options specified but correspondent value is not set because can not be parsed (and we need to rely on error handling to avoid this situation). I believe JVM has mode to ignore incorrect/unknown options so it can lead to unpredictable results (like specified but not set)

cedrichansen replied:
I see. So what you're suggesting then is to move the line 
`extensions->heapExpansionGCRatioThreshold._wasSpecified = true;`
down to a new "else" case below, which would be equivalent to 
`if (OPTION_OK == result)`, is that what you are hoping for?

dmitripivkine replied:
Yes, this is correct. And look like there is no need to extra `FIND_ARG_IN_VMARGS()` call,  `option_set_to_opt_percent()` does it internally

dmitripivkine commented on runtime/gc_vlhgc/MemorySubSpaceTarok.cpp:
```diff
@@ -1199,129 +1340,68 @@ MM_MemorySubSpaceTarok::timeForHeapContract(MM_EnvironmentBase *env, MM_Allocate
-			}
-		} else {
-			/* No need to contract as current free less than max */
+		return 0;
```
Looks like it is easy to keep single return in this function (as recommended), just put remainder if the code to `else`.

dmitripivkine commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -123,6 +174,54 @@ MM_SchedulingDelegate::getInitialTaxationThreshold(MM_EnvironmentVLHGC *env)
 }
 
+void 
+MM_SchedulingDelegate::globalMarkCycleStart(MM_EnvironmentVLHGC *env){
```
```suggestion
MM_SchedulingDelegate::globalMarkCycleStart(MM_EnvironmentVLHGC *env)
{
```

dmitripivkine commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -123,6 +174,54 @@ MM_SchedulingDelegate::getInitialTaxationThreshold(MM_EnvironmentVLHGC *env)
+}
+
+void 
+MM_SchedulingDelegate::calculateGlobalMarkOverhead(MM_EnvironmentVLHGC *env) {
```
```suggestion
MM_SchedulingDelegate::calculateGlobalMarkOverhead(MM_EnvironmentVLHGC *env)
{
```

dmitripivkine commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -123,6 +174,54 @@ MM_SchedulingDelegate::getInitialTaxationThreshold(MM_EnvironmentVLHGC *env)
+
+	/* Total GMP work time, is time taken for all increments + the time we attribute for concurrent GC parts of GMP, and global sweep time. */
+	uint64_t potentialGMPWorkTime =  _globalMarkIncrementsTotalTime + _globalSweepTimeUs + concurrentCostUs;
+	double potentialOverhead = (double)potentialGMPWorkTime/globalMarkIntervalTime;
```
```suggestion
	double potentialOverhead = (double)potentialGMPWorkTime / globalMarkIntervalTime;
```

dmitripivkine commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -123,6 +174,54 @@ MM_SchedulingDelegate::getInitialTaxationThreshold(MM_EnvironmentVLHGC *env)
+		_totalGMPWorkTimeUs = (_historicalPartialGCTime * 1000) * 5;
+	} 
+
+	_globalMarkOverhead = (double)_totalGMPWorkTimeUs/globalMarkIntervalTime;
```
```suggestion
	_globalMarkOverhead = (double)_totalGMPWorkTimeUs / globalMarkIntervalTime;
```

dmitripivkine commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -183,12 +289,31 @@ MM_SchedulingDelegate::globalGarbageCollectCompleted(MM_EnvironmentVLHGC *env, U
+}
+
+void
+MM_SchedulingDelegate::calculatePartialGarbageCollectOverhead(MM_EnvironmentVLHGC *env) {
```
```suggestion
MM_SchedulingDelegate::calculatePartialGarbageCollectOverhead(MM_EnvironmentVLHGC *env)
{
```

dmitripivkine commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -183,12 +289,31 @@ MM_SchedulingDelegate::globalGarbageCollectCompleted(MM_EnvironmentVLHGC *env, U
+	double recentOverhead = (double)(_historicalPartialGCTime * 1000)/_averagePgcInterval;
+	_partialGcOverhead = MM_Math::weightedAverage(_partialGcOverhead, recentOverhead, pgcOverheadHistoricWeight);
+	
+	Trc_MM_SchedulingDelegate_calculatePartialGarbageCollectOverhead(env->getLanguageVMThread(), _partialGcOverhead, _averagePgcInterval/1000, _historicalPartialGCTime);
```
Would you please make this single return function (by inverting logical condition)?

dmitripivkine commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -524,44 +663,195 @@ MM_SchedulingDelegate::calculateEstimatedGlobalBytesToScan() const
+	 * while staying below the specific gc pause time threshold
+	 */
+
+	uint64_t avgPgcTimeUs = _historicalPartialGCTime * 1000;
```
I am ok with this "as-is". However if it helps to improve quality of calculation (and be more natural) we can store `_historicalPartialGCTime` in microseconds and use /1000 where it necessary. This time comes from PGC time converted to milliseconds in `MM_SchedulingDelegate::partialGarbageCollectCompleted()` - line touched in this change set earlier

cedrichansen replied:
Right I agree that In some places it isn't very natural. My original thought was to keep `_historicalPartialGCTime ` in ms (since it is used elsewhere), and convert to us where necessary, but it seems like there are more places where it is used in us now...

I think your idea of doing this in a separate change might be best, that way it can be the main focus on the change, and given proper attention

dmitripivkine commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -524,44 +663,195 @@ MM_SchedulingDelegate::calculateEstimatedGlobalBytesToScan() const
+MM_SchedulingDelegate::predictIntervalBetweenCollections(MM_EnvironmentVLHGC *env, uintptr_t currentEdenSize, intptr_t edenSizeChange, uint64_t pgcAvgIntervalTime)
+{	
+	/* The interval between PGC collections is proportional to eden size. Ex. If eden size doubles, we expect the interval between PGC collections to double as well */
+	double intervalChange = (double)(currentEdenSize + edenSizeChange)/currentEdenSize;
```
```suggestion
	double intervalChange = (double)(currentEdenSize + edenSizeChange) / currentEdenSize;
```

dmitripivkine commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -524,44 +663,195 @@ MM_SchedulingDelegate::calculateEstimatedGlobalBytesToScan() const
+MM_SchedulingDelegate::predictNumberOfCollections(MM_EnvironmentVLHGC *env, uintptr_t currentEdenSize, intptr_t edenSizeChange, uintptr_t freeTenure)
+{
+	/* The number of PGC collections is proportional to how much free tenure will be left after we expand/contract eden */
+	double collectionCountChange = (double)(freeTenure - edenSizeChange)/freeTenure;
```
```suggestion
	double collectionCountChange = (double)(freeTenure - edenSizeChange) / freeTenure;
```

dmitripivkine commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -524,44 +663,195 @@ MM_SchedulingDelegate::calculateEstimatedGlobalBytesToScan() const
+	 * Create a model to determine how pgc time will be afffected by eden size - keeping in mind that _pgcTimeIncreasePerEdenFactor can vary depending on the application
+	 */
+	double edenSizeGb = (double)currentEdenSize / 1000000000;
+	double edenChangeGb = (double)edenSizeChange/ 1000000000;
```
```suggestion
	double edenChangeGb = (double)edenSizeChange / 1000000000;
```

dmitripivkine commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -524,44 +663,195 @@ MM_SchedulingDelegate::calculateEstimatedGlobalBytesToScan() const
+	double edenChangeRatio = (edenChangeGb + edenSizeGb + 1.0) / (edenSizeGb + 1.0);
+
+	/* Use a math workaround for "log base _pgcTimeIncreasePerEdenFactor (edenChangeRatio) "*/
+	double pgcTimeChangeForEdenChange = log(edenChangeRatio) / log(_pgcTimeIncreasePerEdenFactor);
```
This variable name sounds "heavy". `pgcTimeChangeForEden`? `pgcTimeDeltaForEdenChange`? Something else?

cedrichansen replied:
Agreed. I like `pgcTimeDeltaForEdenChange`

dmitripivkine commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -973,65 +1302,317 @@ MM_SchedulingDelegate::calculateAutomaticGMPIntermission(MM_EnvironmentVLHGC *en
+intptr_t
+MM_SchedulingDelegate::moveTowardRecommendedEdenForExpandedHeap(MM_EnvironmentVLHGC *env, double edenChangeSpeed) 
+{
+	Assert_MM_true((edenChangeSpeed <= 1) && (edenChangeSpeed >= 0));
```
Should constants double compared with have double type too? Like `1.` and `0.`?
I guess there are more places we did not pay attention to it

cedrichansen replied:
yes, there are multiple places where I omitted the `.0` (for no particularly good reason).
I'll go through the change set and add `.0` where necessary

dmitripivkine commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -973,65 +1302,317 @@ MM_SchedulingDelegate::calculateAutomaticGMPIntermission(MM_EnvironmentVLHGC *en
+
+	if ((0 == _historicalPartialGCTime) || (0 == _averagePgcInterval)) {
+		/* Until we have collected any information about PGC time, we don't have the data we need to make informed decision about eden size  */
+		return 0;
```
this function is simple enough to be converted to single return function

dmitripivkine commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -973,65 +1302,317 @@ MM_SchedulingDelegate::calculateAutomaticGMPIntermission(MM_EnvironmentVLHGC *en
+		if (globalSweepHappened) {
+			heapFullyExpandedRecommendation = moveTowardRecommendedEdenForExpandedHeap(env, 0.5);
+			heapNotFullyExpandedRecommendation = calculateEdenChangeHeapNotFullyExpanded(env);
+		} else if (0 == _pgcCountSinceGMPEnd % consecutivePGCToChangeEden){
```
```suggestion
		} else if (0 == (_pgcCountSinceGMPEnd % consecutivePGCToChangeEden)){
```

dmitripivkine commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -973,65 +1302,317 @@ MM_SchedulingDelegate::calculateAutomaticGMPIntermission(MM_EnvironmentVLHGC *en
+	
+	/* Convert expectedTimeRatioMinimum/Maximum to 0-100 based for this formula */
+	const double xminpct = _extensions->dnssExpectedRatioMinimum._valueSpecified * 100;
+	const double xmaxpct = _extensions->dnssExpectedRatioMaximum._valueSpecified * 100;
```
double constants... should be `(double)100` or `100.0`

dmitripivkine commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -973,65 +1302,317 @@ MM_SchedulingDelegate::calculateAutomaticGMPIntermission(MM_EnvironmentVLHGC *en
+		 * The overhead logic here will map a low avg pgc time (ie, under targetPauseTimeMs), to a low overhead value (aka, a "better"/more desirable value)
+		 * Ex. Suppose tarokTargetMaxPauseTime == 100ms. 20ms -> 5% (good/desirable), 100ms -> 5% (still good), 110ms -> 6% (should possibly shrink) 500ms -> 80% (bad/undesirable/eden should probably shrink)
+		 */
+		double midpointPct = (xmaxpct + xminpct)/2.0;
```
```suggestion
		double midpointPct = (xmaxpct + xminpct) / 2.0;
```

dmitripivkine commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -973,65 +1302,317 @@ MM_SchedulingDelegate::calculateAutomaticGMPIntermission(MM_EnvironmentVLHGC *en
+		 * 
+		 *  NOTE: This mapped overhead value is later blended with pgc cpu overhead. This means that if the pgc pause time mapping suggested eden contraction, contraction is not guaranteed.
+		 */
+		double slope = (xminpct - xmaxpct) / 20;
```
`20.0`?

cedrichansen replied:
I'll move that to a constant at the top of the file, and add a comment describing what it means/does.

One of the models that was developed to support this mapping, uses 20, and it empirically produces good results. 
The larger this value, the smaller the "penalty" for being over the target pause time becomes smaller, whereas the closer to 0, the penalty becomes massive, which leads to PGC times respecting the pause times very very tighly.

20.0 gives a great balance. If PGC cpu overhead is suffering a lot, and can benefit from going a bit over the pause time target, then it will do so. 20 simply gives a good balance here. 

With a value like 1000 for example, the logic will barely care about the pause time target, and mostly focus on PGC overhead - This is not great, because then target pause time barely does anything.

With a value like 0.1, the pause time target will almost never be surpassed - this is NOT ideal. Certain situations will have very high PGC overhead if we fully respect a very low target pause time

dmitripivkine commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -973,65 +1302,317 @@ MM_SchedulingDelegate::calculateAutomaticGMPIntermission(MM_EnvironmentVLHGC *en
+		 * Suggesting expansion simply because pgc time is small, isn't necessarily a good idea. 
+		 * Instead, if the pgc pause time is still relatively far from target pause time, simply return _partialGcOverhead so that pgc cpu overhead, so that pgc cpu overhead is effectively dictating eden size  
+		 */
+		overhead = OMR_MIN(overhead, _partialGcOverhead * 100);
```
100.0 ?

dmitripivkine commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -973,65 +1302,317 @@ MM_SchedulingDelegate::calculateAutomaticGMPIntermission(MM_EnvironmentVLHGC *en
+	 */
+	double pgcTimeOverhead = mapPgcPauseOverheadToPgcCPUOverhead(env, pgcPauseTimeMs, heapFullyExpanded);
+	double hybridEdenOverheadHundredBased = MM_Math::weightedAverage(overhead * 100, pgcTimeOverhead, pgcCpuOverheadWeight);
+	return hybridEdenOverheadHundredBased / 100;
```
100.0 for two lines

dmitripivkine commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -1157,16 +1724,44 @@ MM_SchedulingDelegate::updateGMPStats(MM_EnvironmentVLHGC *env)
+	double y1 = (double)minimumPgcTime;
+
+	double x2 = (double)getCurrentEdenSizeInBytes(env) / 1000000000;
+	double y2 = (double)_historicalPartialGCTime;
```
both places `1000000000.0`

cedrichansen commented:
@dmitripivkine All PR comments have been addressed (as indicated by "thumbs up" emoji on each comment) as of the latest commit 
with the exception of the following comments:  
- https://github.com/eclipse-openj9/openj9/pull/12043#discussion_r682660003
- https://github.com/eclipse-openj9/openj9/pull/12043#discussion_r683692380

The following comment has been addressed, and I have added some extra notes/explanations to clarify what was done
- https://github.com/eclipse-openj9/openj9/pull/12043#discussion_r683743814

These might require some additional discussion

amicic commented:
Jenkins test sanity all jdk11

amicic commented:
Jenkins test sanity all jdk11

dmitripivkine reviewed (approved):
There is minor issue with parsing of options. Ok, we can fix it later is separate PR
